doc/man/man4/loop.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/loop.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH LOOP 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B loop
\- \fISTREAMS\fP loop-around pseudo-device driver
.\"
.\"
.SH DESCRIPTION
.PP
The
.B loop
psuedo-device multiplexing driver is an implementation of the loop-around driver described in the
.RI \(lq "UNIX\(rg System V Release 4 \- STREAMS Programmer's Guide" ,\(rq
.[
svr4spg
.]
with a number of exceptions and enhancments.  The primary purposes of the
.B loop
pseudo-device driver are: to provide a capability to the
.BR test-streams (8)
conformance and verification test program, for testing passing of STREAMS file descriptors across a
pipe-like driver;
and to provide an example of a pipe-like driver for the STREAM programmer.
To avoid complications for testing, and to serve as a basic example, the driver is as simplistic as
possible.
.PP
The
.B loop
pseudo-device driver has the following behaviour:
.TP 3n
.B "open\fR(2s)"
When a
.B loop
STREAM is opened using
.BR open (2s)
a new
.B loop
STREAM is created.  The newly created STREAM is not associated with any other STREAM.
Issuing the
.B LOOP_SET
input-output control is necessary, to associate the loop STREAM with another one,
before data can be written to or read from the STREAM.  Any attempt to send data on the STREAM
before it is associated with another will result in an
.BR M_ERROR (9)
message being sent upstream causing all further read, write or control operations on the stream to
fail with error
.RB [ ENXIO ].
.TP
.B LOOP_SET
The
.BR LOOP_SET
input-output control described under
.RI \(lq IOCTLS ,\(rq
below, is used to connect the upstream and downstream flow between the requesting STREAM and a
specified STREAM.  Once connected in this fashion, data can be passed in either direction.
.TP
.B "putpmsg\fR(2s), \fPgetpmsg\fR(2s)"
Data is passed in either direction using the
.BR putpmsg (2s)
or
.BR getpmsg (2s)
calls as normal.  The
.B loop
driver coordinates
.I STREAMS
flow control across the driver by explicitly enabling service procedures of the feeding side when
the accepting side is back-enabled.
.TP
.B "close\fR(2s)"
When an STREAM is closed using the
.BR close (2s)
system call, the STREAM is dismantled.
If the upper STREAM remains connected to another STREAM (due to a previous call to
.BR LOOP_SET ),
the connection will be broken and a
.BR M_HANGUP (9)
message will be sent to the other STREAM.
.\"
.\"
.SH USAGE
.PP
The
.B loop
pseudo-device loop-around driver is an implementation of the loop-around driver described in the
.RI \(lq "UNIX\(rg System V Release 4 \- STREAMS Programmer's Guide" ,\(rq
.[
svr4spg
.]
with a number of considerations as follows:
.IP \(bu \w'\(bu\(em'u
The
.I SVR4
driver does not implement input-output controls other than
.BR LOOP_SET .
In contrast,
.B loop
implements a set of input-output controls compatible with the
loop-around driver for use in testing.
.\"
.\"
.SH IOCTLS
.PP
.TP
.B LOOP_SET
.\"
.\"
.SH NOTICES
.PP
The
.B loop
driver is used by the
.BR test-streams (8)
conformance and verification test program to test passing of file descriptors across a
pipe-like driver.  Much of its behaviour is specific to testing and cannot be altered without
impacting the ability of the
.BR test-streams (8)
conformance and verification test program results.
.\"
.\"
.SH IMPLEMENTATION
.PP
.B loop
is implemented as described in the
.IR "SVR4 SPG" .
.[
svr4spg
.]
.\"
.\"
.\".SH EXAMPLES
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/sloop
The external filesystem
.I STREAMS
loop device.
.TP
.BR /dev/streams/loop ", " /devices/loop
The
.BR specfs (5)
shadow special filesystem
.I STREAMS
loop device.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR mux (4),
.BR nullmod (4),
.BR test-streams (8).
.\"
.\"
.SH BUGS
.PP
.B loop
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B @PACKAGE_TITLE@
.[
lfs
.]
provides the
.B loop
pseudo-device driver for testing
with the following portability considerations:
.IP \(em \w'\(em\(em'u
Although this driver is described in the
.IR "SVR4 SPG"
.[
svr4spg
.]
as a driver example, most
.I STREAMS
implementations do not actually provide such a driver.
.IP \(em
.B loop
is provided primarily for use by the
.BR test-streams (8)
conformance and verification test program, and as an example of a pipe-like driver with flow
control supported across the driver.  It should not be used by portable
.I STREAMS
modules, driver or applications programs.
.PP
See
.IR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
SVR4.
.[
svr4spg
.]
Conformance is verified using the
.BR test-loop (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
.B loop
is described as an example loop-around driver in the
.IR "SVR4 SPG" .
.[
svr4spg
.]
.\"
.\"
.[
$LIST$
.]
.TI



NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/loop.c
/*****************************************************************************

 @(#) File: src/drivers/loop.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/loop.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This file contains a classic loop driver for SVR 4.2 MP STREAMS.  The loop driver is a general
 *  purpose STREAMS-based pipe-like driver that can be used for connecting two upper-level STREAMS
 *  together to form a structure similar to a STREAMS-based Pipe, but which has an additional driver
 *  module beneath each STREAM head.  Upper level STREAMS are linked together using ioctl()
 *  commands contained in <sys/loop.h>.
 */
#ifdef NEED_LINUX_AUTOCONF_H
#include NEED_LINUX_AUTOCONF_H
#endif
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/sched.h>
#include <linux/interrupt.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>
#include <sys/loop.h>

#include "sys/config.h"

#define LOOP_DESCRIP	"SVR 4.2 STREAMS Null Stream (LOOP) Driver"
#define LOOP_EXTRA	"Part of UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define LOOP_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define LOOP_REVISION	"OpenSS7 src/drivers/loop.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define LOOP_DEVICE	"SVR 4.2 MP STREAMS Null Stream (LOOP) Device"
#define LOOP_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define LOOP_LICENSE	"GPL"
#define LOOP_BANNER	LOOP_DESCRIP	"\n" \
			LOOP_EXTRA	"\n" \
			LOOP_COPYRIGHT	"\n" \
			LOOP_REVISION	"\n" \
			LOOP_DEVICE	"\n" \
			LOOP_CONTACT	"\n"
#define LOOP_SPLASH	LOOP_DEVICE	" - " \
			LOOP_REVISION	"\n"

#ifdef CONFIG_STREAMS_LOOP_MODULE
MODULE_AUTHOR(LOOP_CONTACT);
MODULE_DESCRIPTION(LOOP_DESCRIP);
MODULE_SUPPORTED_DEVICE(LOOP_DEVICE);
MODULE_LICENSE(LOOP_LICENSE);
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_LOOP_MODULE */

#ifndef CONFIG_STREAMS_LOOP_NAME
#define CONFIG_STREAMS_LOOP_NAME "loop"
#endif
#ifndef CONFIG_STREAMS_LOOP_MODID
#error CONFIG_STREAMS_LOOP_MODID must be defined.
#endif
#ifndef CONFIG_STREAMS_LOOP_MAJOR
#error CONFIG_STREAMS_LOOP_MAJOR must be defined.
#endif

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-loop");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_LOOP_MODULE
static
#endif
modID_t modid = CONFIG_STREAMS_LOOP_MODID;

#ifdef CONFIG_STREAMS_LOOP_MODULE
#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module id number for LOOP driver. (0 for auto allocation)");
#endif				/* CONFIG_STREAMS_LOOP_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_LOOP_MODID));
MODULE_ALIAS("streams-driver-loop");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_LOOP_MODULE
static
#endif
major_t major = CONFIG_STREAMS_LOOP_MAJOR;

#ifdef CONFIG_STREAMS_LOOP_MODULE
#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for LOOP driver. (0 for auto allocation)");
#endif				/* CONFIG_STREAMS_LOOP_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_CLONE_MAJOR) "-" __stringify(CONFIG_STREAMS_LOOP_MAJOR));
MODULE_ALIAS("/dev/loop.1");
MODULE_ALIAS("/dev/loop.2");
MODULE_ALIAS("/dev/loop_clone");
//MODULE_ALIAS("devname:loop_clone");
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_LOOP_MAJOR));
MODULE_ALIAS("/dev/streams/loop");
MODULE_ALIAS("/dev/streams/loop/*");
MODULE_ALIAS("/dev/streams/clone/loop");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

STATIC struct module_info loop_minfo = {
	.mi_idnum = CONFIG_STREAMS_LOOP_MODID,
	.mi_idname = CONFIG_STREAMS_LOOP_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat loop_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat loop_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

typedef struct loop {
	struct loop *next;		/* list linkage */
	struct loop **prev;		/* list linkage */
	struct loop *other;		/* other stream */
	dev_t dev;			/* device number */
	queue_t *rq;			/* this rd queue */
	queue_t *wq;			/* this wr queue */
} loop_t;

STATIC struct loop *loop_opens = NULL;
#if	defined DEFINE_SPINLOCK
STATIC DEFINE_SPINLOCK(loop_lock);
#elif	defined __SPIN_LOCK_UNLOCKED
STATIC spinlock_t loop_lock = __SPIN_LOCK_UNLOCKED(loop_lock);
#elif	defined SPIN_LOCK_UNLOCKED
STATIC spinlock_t loop_lock = SPIN_LOCK_UNLOCKED;
#else
#error cannot initialize spin locks
#endif

/*
 *  Locking
 */
#if defined CONFIG_STREAMS_NOIRQ || defined _TEST

#define spin_lock_str(__lkp, __flags) \
	do { (void)__flags; spin_lock_bh(__lkp); } while (0)
#define spin_unlock_str(__lkp, __flags) \
	do { (void)__flags; spin_unlock_bh(__lkp); } while (0)

#else

#define spin_lock_str(__lkp, __flags) \
	spin_lock_irqsave(__lkp, __flags)
#define spin_unlock_str(__lkp, __flags) \
	spin_unlock_irqrestore(__lkp, __flags)

#endif

STATIC streamscall int
loop_wput(queue_t *q, mblk_t *mp)
{
	struct loop *p = q->q_ptr, *o;
	unsigned long flags;
	int err;

	switch (mp->b_datap->db_type) {
	case M_IOCTL:
	{
		union ioctypes *ioc = (typeof(ioc)) mp->b_rptr;

		switch (ioc->iocblk.ioc_cmd) {
		case LOOP_SET:
		{
			dev_t dev;

			if (ioc->iocblk.ioc_count != sizeof(dev)) {
				err = -EINVAL;
				break;
			}
			if (mp->b_cont == NULL) {
				err = -EINVAL;
				break;
			}
			dev = *(dev_t *) mp->b_cont->b_rptr;
			spin_lock_str(&loop_lock, flags);
			for (o = loop_opens; o; o = o->next)
				if (o->dev == dev)
					break;
			if (o == NULL) {
				spin_unlock_str(&loop_lock, flags);
				err = -ENXIO;
				break;
			}
			if (p->other || o->other) {
				spin_unlock_str(&loop_lock, flags);
				err = -EBUSY;
				break;
			}
			p->other = o;
			o->other = p;
			enableok(o->wq);
			enableok(p->wq);
			qenable(o->wq);
			qenable(p->wq);
			spin_unlock_str(&loop_lock, flags);
			err = 0;
			break;
		}
		default:
			err = -EINVAL;
			break;
		}
		if (err == 0) {
			mp->b_datap->db_type = M_IOCACK;
			ioc->iocblk.ioc_count = 0;
			ioc->iocblk.ioc_rval = 0;
			ioc->iocblk.ioc_error = 0;
		} else {
			mp->b_datap->db_type = M_IOCNAK;
			ioc->iocblk.ioc_count = 0;
			ioc->iocblk.ioc_rval = -1;
			ioc->iocblk.ioc_error = -err;
		}
		qreply(q, mp);
		break;
	}
	case M_FLUSH:
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
		}
		spin_lock_str(&loop_lock, flags);
		if (p->other) {
			queue_t *rq;

			rq = p->other->rq;
			spin_unlock_str(&loop_lock, flags);
			/* switch sense of flush flags */
			switch (mp->b_rptr[0] & (FLUSHW | FLUSHR | FLUSHRW)) {
			case FLUSHR:
				mp->b_rptr[0] = (mp->b_rptr[0] & ~FLUSHR) | FLUSHW;
				break;
			case FLUSHW:
				mp->b_rptr[0] = (mp->b_rptr[0] & ~FLUSHW) | FLUSHR;
				break;
			}
			putnext(rq, mp);
		} else {
			spin_unlock_str(&loop_lock, flags);
			if (mp->b_rptr[0] & FLUSHR) {
				if (mp->b_rptr[0] & FLUSHBAND)
					flushband(RD(q), mp->b_rptr[1], FLUSHDATA);
				else
					flushq(RD(q), FLUSHDATA);
				mp->b_rptr[0] &= ~FLUSHW;
				qreply(q, mp);
			} else {
				freemsg(mp);
			}
		}
		break;
	default:
		spin_lock_str(&loop_lock, flags);
		if (p->other) {
			if (!q->q_first) {
				queue_t *rq;

				rq = p->other->rq;
				spin_unlock_str(&loop_lock, flags);
				if (rq) {
					if (mp->b_datap->db_type >= QPCTL
					    || bcanputnext(rq, mp->b_band)) {
						putnext(rq, mp);
						break;
					}
				}
			} else
				spin_unlock_str(&loop_lock, flags);
			putq(q, mp);
		} else {
			spin_unlock_str(&loop_lock, flags);
			freemsg(mp);
			putnextctl2(OTHERQ(q), M_ERROR, ENXIO, ENXIO);
		}
		break;
	}
	return (0);
}

/*
 *  The driver only puts messages on its write queue when they fail flow control on the queue beyond
 *  the other read queue.  When flow control subsides, the read service procedure on the othe STREAM
 *  will explicitly enable us.
 */
STATIC streamscall int
loop_wsrv(queue_t *q)
{
	struct loop *p = q->q_ptr;
	unsigned long flags;
	queue_t *rq = NULL;
	mblk_t *mp;

	spin_lock_str(&loop_lock, flags);
	if (p->other)
		rq = p->other->rq;
	spin_unlock_str(&loop_lock, flags);

	if (rq) {
		while ((mp = getq(q))) {
			if (mp->b_datap->db_type >= QPCTL || bcanputnext(rq, mp->b_band)) {
				putnext(rq, mp);
				continue;
			}
			putbq(q, mp);
			break;
		}
	} else
		flushq(q, FLUSHALL);

	return (0);
}

/*
 *  The driver never places messages on the read queue, therefore, the read service procedure is
 *  only invoked when backenabled from upstream.  Back enabling is due to flow control that caused a
 *  message to tbe put (back) to the other STREAM's write queue so the driver needs to explicitly
 *  enable that write queue now.
 */
STATIC streamscall int
loop_rsrv(queue_t *q)
{
	struct loop *p = q->q_ptr;
	unsigned long flags;

	spin_lock_str(&loop_lock, flags);
	if (p->other && p->other->wq)
		qenable(p->other->wq);
	spin_unlock_str(&loop_lock, flags);
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
STATIC streamscall int
loop_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	struct loop *p, **pp = &loop_opens;
	major_t cmajor = getmajor(*devp);
	minor_t cminor = getminor(*devp);
	unsigned long flags;

	if (q->q_ptr != NULL) {
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q)->q_next) {
		return (ENXIO);	/* can't open as module */
	}
	if (!(p = kmem_alloc(sizeof(*p), KM_NOSLEEP))) {	/* we could sleep */
		return (ENOMEM);	/* no memory */
	}
	bzero(p, sizeof(*p));
	switch (sflag) {
	case CLONEOPEN:
		if (cminor < 1)
			cminor = 1;
	case DRVOPEN:
	{
		major_t dmajor = cmajor;

		if (cminor < 1) {
			return (ENXIO);
		}
		spin_lock_str(&loop_lock, flags);
		for (; *pp && (dmajor = getmajor((*pp)->dev)) < cmajor; pp = &(*pp)->next) ;
		for (; *pp && dmajor == getmajor((*pp)->dev) &&
		     getminor(makedevice(cmajor, cminor)) != 0; pp = &(*pp)->next) {
			minor_t dminor = getminor((*pp)->dev);

			if (cminor < dminor)
				break;
			if (cminor == dminor) {
				if (sflag == CLONEOPEN)
					cminor++;
				else {
					spin_unlock_str(&loop_lock, flags);
					kmem_free(p, sizeof(*p));
					pswerr(("%s: stream already open!\n", __FUNCTION__));
					return (EIO);	/* bad error */
				}
			}
		}
		if (getminor(makedevice(cmajor, cminor)) == 0) {	/* no minors left */
			spin_unlock_str(&loop_lock, flags);
			kmem_free(p, sizeof(*p));
			return (EBUSY);	/* no minors left */
		}
		p->dev = *devp = makedevice(cmajor, cminor);
		p->rq = q;
		p->wq = WR(q);
		noenable(p->wq);
		if ((p->next = *pp))
			p->next->prev = &p->next;
		p->prev = pp;
		*pp = p;
		q->q_ptr = WR(q)->q_ptr = p;
		spin_unlock_str(&loop_lock, flags);
		qprocson(q);
		return (0);
	}
	}
	pswerr(("%s: bad sflag %d\n", __FUNCTION__, sflag));
	return (ENXIO);
}

STATIC streamscall int
loop_close(queue_t *q, int oflag, cred_t *crp)
{
	struct loop *p;
	unsigned long flags;

	if ((p = q->q_ptr) == NULL) {
		pswerr(("%s: already closed\n", __FUNCTION__));
		return (0);	/* already closed */
	}
	qprocsoff(q);
	spin_lock_str(&loop_lock, flags);
	if (p->other && p->other->rq)
		putnextctl(p->other->rq, M_HANGUP);
	if ((*(p->prev) = p->next))
		p->next->prev = p->prev;
	p->next = NULL;
	p->prev = &p->next;
	p->other = NULL;
	p->dev = 0;
	p->rq = NULL;
	p->wq = NULL;
	q->q_ptr = WR(q)->q_ptr = NULL;
	spin_unlock_str(&loop_lock, flags);
	return (0);
}

STATIC struct qinit loop_rqinit = {
	.qi_srvp = loop_rsrv,
	.qi_qopen = loop_open,
	.qi_qclose = loop_close,
	.qi_minfo = &loop_minfo,
	.qi_mstat = &loop_rstat,
};

STATIC struct qinit loop_wqinit = {
	.qi_putp = loop_wput,
	.qi_srvp = loop_wsrv,
	.qi_minfo = &loop_minfo,
	.qi_mstat = &loop_wstat,
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
STATIC
#endif
struct streamtab loopinfo = {
	.st_rdinit = &loop_rqinit,
	.st_wrinit = &loop_wqinit,
};

STATIC struct cdevsw loop_cdev = {
	.d_name = CONFIG_STREAMS_LOOP_NAME,
	.d_str = &loopinfo,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
STATIC
#endif
int __init
loopinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_LOOP_MODULE
	printk(KERN_INFO LOOP_BANNER);
#else
	printk(KERN_INFO LOOP_SPLASH);
#endif
	loop_minfo.mi_idnum = modid;
	if ((err = register_strdev(&loop_cdev, major)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
STATIC
#endif
void __exit
loopexit(void)
{
	unregister_strdev(&loop_cdev, major);
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
module_init(loopinit);
module_exit(loopexit);
#endif
